projects/stackoverflow/src/stackoverflow/DataExtractor.java

```diff
@@ -0,0 +1,178 @@
+package stackoverflow;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.util.Iterator;
+
+import javax.xml.stream.XMLEventFactory;
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.XMLEventWriter;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.events.Attribute;
+import javax.xml.stream.events.XMLEvent;
+
+public class DataExtractor {
+
+	// paths of the files that we read from and write to
+	private String inputFilePath;
+	private String outputFilePath;
+
+	// the name of the row tag.
+	private static final String ROW = "row";
+	// the name of the posts tag
+	private static final String POSTS = "posts";
+	// some constant strings for the relevant attributes's names in the row tag
+	private static final String ID = "Id";
+	private static final String Post_TYPE_ID = "PostTypeId";
+	private static final String CREATION_DATE = "CreationDate";
+	private static final String OWNER_USER_ID = "OwnerUserId";
+	private static final String TAGS = "Tags";
+
+	public void extractData() {
+		// deffine those variables here so they can be cleaned in finally block
+		XMLEventReader eventReader = null;
```

Comment:
You can use try-with-resources to avoid having to initialize and then close:
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Note that you can have a single try with both resources.

Reply:
Unfortunately, it seems that try-with-resources won't help us here. The classes XMLEventReader and XMLEventWriter doesn't seem to implement java.lang.AutoCloseable interface, so the try-with-resources can auto close them.

Reply:
Ok, got it.